File: examples/ocaml/x16.ml
(*
        plshade demo, using color fill.

        Maurice LeBrun
        IFS, University of Texas at Austin
        20 Mar 1994
*)

open Plplot

let colorbar ?color ?contour values =
  (* Smaller text *)
  plschr 0.0 0.75;
  (* Small ticks on the vertical axis *)
  plsmaj 0.0 0.5;
  plsmin 0.0 0.5;

  let axis =
    [
      `frame0;
      `frame1;
      `vertical_label;
      `unconventional_label;
      `major_ticks;
    ]
  in
  let shade = Plot.shade_colorbar ~custom:true ~axis values in
  let pos = Plot.viewport_pos ~inside:false 0.005 0.0 in
  Plot.plot [
    Plot.colorbar ?color ?contour ~orient:(`top (0.0375, 0.875)) ~label:[`bottom "Magnitude"] ~pos shade;
  ];

  (* Reset text and tick sizes *)
  plschr 0.0 1.0;
  plsmaj 0.0 1.0;
  plsmin 0.0 1.0

let pi = atan 1.0 *. 4.0

(* Fundamental settings.  See notes[] for more info. *)

let ns = 20             (* Default number of shade levels *)
let nx = 35             (* Default number of data points in x *)
let ny = 46             (* Default number of data points in y *)
let exclude = false     (* By default do not plot a page illustrating
                           exclusion. *)

(* polar plot data *)
let perimeterpts = 100

(* Transformation function *)
let mypltr x y tr =
  tr.(0) *. x +. tr.(1) *. y +. tr.(2),
  tr.(3) *. x +. tr.(4) *. y +. tr.(5)

let zdefined x y =
  let z = sqrt (x *. x +. y *. y) in
  if z < 0.4 || z > 0.6 then 1 else 0

(*--------------------------------------------------------------------------*\
 * f2mnmx
 *
 * Returns min & max of input 2d array.
\*--------------------------------------------------------------------------*)
let f2mnmx f =
  let fmax = ref f.(0).(0) in
  let fmin = ref f.(0).(0) in
  for i = 0 to Array.length f - 1 do
    for j = 0 to Array.length f.(i) - 1 do
      fmax := max !fmax f.(i).(j);
      fmin := min !fmin f.(i).(j);
    done
  done;
  !fmin, !fmax

(*--------------------------------------------------------------------------*\
 * Does several shade plots using different coordinate mappings.
\*--------------------------------------------------------------------------*)
let () =
  let fill_width = 2.0 in
  let cont_color = 0 in
  let cont_width = 0.0 in

  (* Parse and process command line arguments *)
  plparseopts Sys.argv [PL_PARSE_FULL];

  (* Load color palettes *)
  plspal0 "cmap0_black_on_white.pal";
  plspal1 "cmap1_gray.pal" true;

  (* Reduce colors in cmap 0 so that cmap 1 is useful on a 16-color display *)
  plscmap0n 3;

  (* Initialize plplot *)
  plinit ();

  (* Set up transformation function *)
  let tr =
    [|
      2.0 /. float_of_int (nx - 1); 0.0; -1.0;
      0.0; 2.0 /. float_of_int (ny-1); -1.0;
    |]
  in

  (* Set up data arrays *)
  let z = Array.make_matrix nx ny 0.0 in
  let w = Array.make_matrix nx ny 0.0 in
  for i = 0 to nx - 1 do
    let x = float_of_int (i - (nx / 2)) /. float_of_int (nx / 2) in
    for j = 0 to ny - 1 do
      let y = float_of_int (j - (ny / 2)) /. float_of_int (ny / 2) -. 1.0 in
      z.(i).(j) <- -. sin (7.0 *. x) *. cos (7.0 *. y) +. x *. x -. y *. y;
      w.(i).(j) <- -. cos (7.0 *. x) *. sin (7.0 *. y) +. 2.0 *. x *. y;
    done
  done;

  let zmin, zmax = f2mnmx z in
  let clevel =
    Array.init ns (
      fun i ->
        zmin +. (zmax -. zmin) *. (float_of_int i +. 0.5) /. float_of_int ns
    )
  in
  let shedge =
    Array.init (ns + 1) (
      fun i ->
        zmin +. (zmax -. zmin) *. float_of_int i /. float_of_int ns
    )
  in

  (* Set up coordinate grids *)
  let xg1 = Array.make nx 0.0 in
  let yg1 = Array.make ny 0.0 in
  let xg2 = Array.make_matrix nx ny 0.0 in
  let yg2 = Array.make_matrix nx ny 0.0 in

  for i = 0 to nx - 1 do
    for j = 0 to ny - 1 do
      let x, y = mypltr (float_of_int i) (float_of_int j) tr in

      let argx = x *. pi /. 2.0 in
      let argy = y *. pi /. 2.0 in
      let distort = 0.4 in

      xg1.(i) <- x +. distort *. cos argx;
      yg1.(j) <- y -. distort *. cos argy;

      xg2.(i).(j) <- x +. distort *. cos argx *. cos argy;
      yg2.(i).(j) <- y -. distort *. cos argx *. cos argy;
    done
  done;

  (* Plot using identity transform *)
  pladv 0;
  plvpor 0.1 0.9 0.1 0.9;
  plwind (-1.0) 1.0 (-1.0) 1.0;

  plpsty 0;

  plshades z (-1.0) 1.0 (-1.0) 1.0 shedge fill_width cont_color cont_width true;

  colorbar shedge;

  plcol0 1;
  plbox "bcnst" 0.0 0 "bcnstv" 0.0 0;
  plcol0 2;
  pllab "distance" "altitude" "Bogon density";

  (* Plot using 1d coordinate transform *)

  (* Load color palettes *)
  plspal0 "cmap0_black_on_white.pal";
  plspal1 "cmap1_blue_yellow.pal" true;

  pladv 0;
  plvpor 0.1 0.9 0.1 0.9;
  plwind (-1.0) 1.0 (-1.0) 1.0;

  plpsty 0;

  plset_pltr (pltr1 xg1 yg1);
  plshades z (-1.0) 1.0 (-1.0) 1.0 shedge fill_width cont_color cont_width true;

  colorbar shedge;

  plcol0 1;
  plbox "bcnst" 0.0 0 "bcnstv" 0.0 0;
  plcol0 2;
  pllab "distance" "altitude" "Bogon density";

  (* Plot using 2d coordinate transform *)

  (* Load color palettes *)
  plspal0 "cmap0_black_on_white.pal";
  plspal1 "cmap1_blue_red.pal" true;

  pladv 0;
  plvpor 0.1 0.9 0.1 0.9;
  plwind (-1.0) 1.0 (-1.0) 1.0;

  plpsty 0;

  plset_pltr (pltr2 xg2 yg2);
  plshades
    z (-1.0) 1.0 (-1.0) 1.0 shedge fill_width cont_color cont_width false;

  colorbar shedge;

  plcol0 1;
  plbox "bcnst" 0.0 0 "bcnstv" 0.0 0;
  plcol0 2;
  plcont w 1 nx 1 ny clevel;

  pllab "distance" "altitude" "Bogon density, with streamlines";

  (* Plot using 2d coordinate transform *)

  (* Load color palettes *)
  plspal0 "";
  plspal1 "" true;

  pladv 0;
  plvpor 0.1 0.9 0.1 0.9;
  plwind (-1.0) 1.0 (-1.0) 1.0;

  plpsty 0;

  plshades z (-1.0) 1.0 (-1.) 1.0 shedge fill_width 2 3.0 false;

  colorbar ~color:(`index 2) ~contour:(`index 2, 3.0) shedge;

  plcol0 1;
  plbox "bcnst" 0.0 0 "bcnstv" 0.0 0;
  plcol0 2;

  pllab "distance" "altitude" "Bogon density";

  (* Note this exclusion API will probably change. *)

  (* Plot using 2d coordinate transform and exclusion*)
  if exclude then (
    pladv 0;
    plvpor 0.1 0.9 0.1 0.9;
    plwind (-1.0) 1.0 (-1.0) 1.0;

    plpsty 0;

    plset_defined zdefined;
    plshades
      z (-1.0) 1.0 (-1.0) 1.0 shedge fill_width cont_color cont_width false;

    colorbar shedge;
    plunset_defined ();

    plcol0 1;
    plbox "bcnst" 0.0 0 "bcnstv" 0.0 0;

    pllab "distance" "altitude" "Bogon density with exclusion";
  );
  (* Example with polar coordinates. *)

  (* Load colour palettes*)
  plspal0 "cmap0_black_on_white.pal";
  plspal1 "cmap1_gray.pal" true;

  pladv 0;
  plvpor 0.1 0.9 0.1 0.9;
  plwind (-1.0) 1.0 (-1.0) 1.0;
  plpsty 0;

  (* Build new coordinate matrices. *)
  for i = 0 to nx - 1 do
    let r = float_of_int i /. float_of_int (nx - 1) in
    for j = 0 to ny - 1 do
      let t = (2.0 *. pi /. (float_of_int ny -. 1.0)) *. float_of_int j in
      xg2.(i).(j) <- r *. cos t;
      yg2.(i).(j) <- r *. sin t;
      z.(i).(j) <- exp (~-.r *. r) *. cos (5.0 *. pi *. r) *. cos (5.0 *. t);
    done
  done;

  (* Need a new shedge to go along with the new data set. *)
  let zmin, zmax = f2mnmx z in

  let shedge =
    Array.init (ns + 1) (
      fun i ->
        zmin +. (zmax -. zmin) *. float_of_int i /. float_of_int ns
    )
  in

  (*  Now we can shade the interior region. *)
  plshades
    z (-1.0) 1.0 (-1.0) 1.0 shedge fill_width cont_color cont_width false;

  colorbar shedge;

  (* Now we can draw the perimeter.  (If do before, shade stuff may overlap.) *)
  let px = Array.make perimeterpts 0.0 in
  let py = Array.make perimeterpts 0.0 in
  for i = 0 to perimeterpts - 1 do
    let t = (2.0 *. pi /. float_of_int (perimeterpts - 1)) *. float_of_int i in
    px.(i) <- cos t;
    py.(i) <- sin t;
  done;
  plcol0 1;
  plline px py;

  (* And label the plot.*)
  plcol0 2;
  pllab "" "" "Tokamak Bogon Instability";

  (* Clean up *)
  plend ();
  ()

File: examples/python/x16.py
#  Copyright (C) 2001-2016 Alan W. Irwin

#  plshade demo, using color fill.
#
#  This file is part of PLplot.
#
#  PLplot is free software; you can redistribute it and/or modify
#  it under the terms of the GNU Library General Public License as published
#  by the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  PLplot is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU Library General Public License for more details.
#
#  You should have received a copy of the GNU Library General Public License
#  along with PLplot; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#

from numpy import *

NS = 20
NX = 35
NY = 46
PERIMETERPTS = 100
NUM_AXES = 1
NUM_LABELS = 1

XSPA = 2./(NX-1)
YSPA = 2./(NY-1)

tr = array((XSPA, 0.0, -1.0, 0.0, YSPA, -1.0))

def mypltr(x, y, data):
    result0 = data[0] * x + data[1] * y + data[2]
    result1 = data[3] * x + data[4] * y + data[5]
    return array((result0, result1))

def main(w):

    fill_width = 2.
    cont_color = 0
    cont_width = 0.

    n_axis_opts  = NUM_AXES
    axis_opts = zeros(NUM_AXES,"S100")
    axis_opts[0] = "bcvtm"
    num_values = zeros(NUM_AXES,"int");
    values = reshape(zeros(NUM_AXES*(NS+1)),[NUM_AXES,NS+1])
    axis_ticks = zeros(NUM_AXES)
    axis_subticks = zeros(NUM_AXES,"int")

    n_labels = NUM_LABELS
    label_opts = zeros(NUM_LABELS,"int")
    labels = zeros(NUM_LABELS,"S100")
    label_opts[0] = w.PL_COLORBAR_LABEL_BOTTOM
    labels[0] = "Magnitude"

    # Set up data array

    x = (arange(NX) - (NX/2)) / float(NX/2)
    x.shape = (-1,1)
    y = (arange(NY) - (NY/2)) / float(NY/2) - 1.
    zz = -sin(7.*x) * cos(7.*y) + x*x - y*y
    ww = -cos(7.*x) * sin(7.*y) + 2.*x*y

    zmin = min(zz.flat)
    zmax = max(zz.flat)

    clevel = zmin + (zmax - zmin) * (arange(NS)+0.5)/NS
    shedge = zmin + (zmax - zmin) * (arange(NS+1))/NS

    # Build the identity transformation between grid and world coordinates
    # using mypltr.
    # Note that *for the given* tr, mypltr(i,j,tr)[0] is only a function of i
    # and mypltr(i,j,tr)[1] is only function of j.
    xg0 = mypltr(arange(NX),0,tr)[0]
    yg0 = mypltr(0,arange(NY),tr)[1]

    # Build the 1-d coord arrays.

    distort = .4

    xx =  xg0
    xg1 = xx + distort * cos( .5 * pi * xx )
    yy =  yg0
    yg1 = yy - distort * cos( .5 * pi * yy )

    # Build the 2-d coord arrays.
    xx.shape = (-1,1)
    xg2 = xx + distort*cos((0.5*pi)*xx)*cos((0.5*pi)*yy)
    yg2 = yy - distort*cos((0.5*pi)*xx)*cos((0.5*pi)*yy)

    # Plot using identity transform

    # Load first-page colour palettes
    w.plspal0("cmap0_black_on_white.pal")
    w.plspal1("cmap1_gray.pal",1)
    w.plscmap0n(3)

    w.pladv(0)
    w.plvpor(0.1, 0.9, 0.1, 0.9)
    w.plwind(-1.0, 1.0, -1.0, 1.0)
    w.plpsty(0)

#   Note another alternative to produce the identical result in a different way
#   is to use the command:
#   w.plshades(zz, -1.0, 1.0, -1.0, 1.0, shedge, fill_width, 1)
#   The above command works because xmin, xmax, ymin, and ymax do an effective
#   linear transformation on the index ranges.  (We could have dropped
#   xmin, xmax, ymin, ymax since the defaults are -1.0, 1.0, -1.0, 1.0, but
#   for pedagogical reasons we left them in.)  The alternative below using
#   mypltr does the exact same linear transformation because of the way that
#   the tr array has been defined.  Note that when pltr and pltr_data are
#   defined, xmin, xmax, ymin, ymax are completely ignored so we can drop
#   them from the argument list.
    w.plshades(zz, shedge, fill_width, 1, mypltr, tr)

    # Smaller text
    w.plschr(0.0, 0.75 )
    # Small ticks on the vertical axis
    w.plsmaj( 0.0, 0.5 )
    w.plsmin( 0.0, 0.5 )

    num_values[0] = NS + 1
    values[0] = shedge
    (colorbar_width, colorbar_height) = w.plcolorbar ( w.PL_COLORBAR_SHADE | w.PL_COLORBAR_SHADE_LABEL, 0, 0.005, 0.0, 0.0375, 0.875, 0, 1, 1, 0.0, 0.0, cont_color, cont_width, label_opts, labels, axis_opts, axis_ticks, axis_subticks, num_values, values)

    # Reset text and tick sizes
    w.plschr( 0.0, 1.0 )
    w.plsmaj( 0.0, 1.0 )
    w.plsmin( 0.0, 1.0 )

    w.plcol0(1)
    w.plbox( "bcnst", 0., 0, "bcnstv", 0., 0 )
    w.plcol0(2)

    w.pllab( "distance", "altitude", "Bogon density" )

    # Plot using 1d coordinate transform

    w.plspal0("cmap0_black_on_white.pal")
    w.plspal1("cmap1_blue_yellow.pal",1)
    w.plscmap0n(3)

    w.pladv(0)
    w.plvpor(0.1, 0.9, 0.1, 0.9)
    w.plwind(-1.0, 1.0, -1.0, 1.0)
    w.plpsty(0)

    w.plshades(zz, shedge, fill_width, 1, w.pltr1, xg1, yg1)

    # Smaller text
    w.plschr(0.0, 0.75 )
    # Small ticks on the vertical axis
    w.plsmaj( 0.0, 0.5 )
    w.plsmin( 0.0, 0.5 )

    num_values[0] = NS + 1
    values[0] = shedge
    (colorbar_width, colorbar_height) = w.plcolorbar ( w.PL_COLORBAR_SHADE | w.PL_COLORBAR_SHADE_LABEL, 0, 0.005, 0.0, 0.0375, 0.875, 0, 1, 1, 0.0, 0.0, cont_color, cont_width, label_opts, labels, axis_opts, axis_ticks, axis_subticks, num_values, values)

    # Reset text and tick sizes
    w.plschr( 0.0, 1.0 )
    w.plsmaj( 0.0, 1.0 )
    w.plsmin( 0.0, 1.0 )

    w.plcol0(1)
    w.plbox( "bcnst", 0.0, 0, "bcnstv", 0.0, 0 )
    w.plcol0(2)

    w.pllab( "distance", "altitude", "Bogon density" )

    # Plot using 2d coordinate transform

    w.plspal0("cmap0_black_on_white.pal")
    w.plspal1("cmap1_blue_red.pal",1)
    w.plscmap0n(3)

    w.pladv(0)
    w.plvpor(0.1, 0.9, 0.1, 0.9)
    w.plwind(-1.0, 1.0, -1.0, 1.0)
    w.plpsty(0)

    w.plshades(zz, shedge, fill_width, 0, w.pltr2, xg2, yg2)

    # Smaller text
    w.plschr(0.0, 0.75 )
    # Small ticks on the vertical axis
    w.plsmaj( 0.0, 0.5 )
    w.plsmin( 0.0, 0.5 )

    num_values[0] = NS + 1
    values[0] = shedge
    (colorbar_width, colorbar_height) = w.plcolorbar ( w.PL_COLORBAR_SHADE | w.PL_COLORBAR_SHADE_LABEL, 0, 0.005, 0.0, 0.0375, 0.875, 0, 1, 1, 0.0, 0.0, cont_color, cont_width, label_opts, labels, axis_opts, axis_ticks, axis_subticks, num_values, values)

    # Reset text and tick sizes
    w.plschr( 0.0, 1.0 )
    w.plsmaj( 0.0, 1.0 )
    w.plsmin( 0.0, 1.0 )

    w.plcol0(1)
    w.plbox( "bcnst", 0.0, 0, "bcnstv", 0.0, 0 )
    w.plcol0(2)

    w.plcont(ww, clevel, w.pltr2, xg2, yg2)
    w.pllab( "distance", "altitude", "Bogon density, with streamlines" )

    # Plot using 2d coordinate transform (with contours generated by w.plshades)

    w.plspal0("")
    w.plspal1("",1)
    w.plscmap0n(3)

    w.pladv(0)
    w.plvpor(0.1, 0.9, 0.1, 0.9)
    w.plwind(-1.0, 1.0, -1.0, 1.0)
    w.plpsty(0)

    # Note default cont_color and cont_width are zero so that no contours
    # are done with other calls to w.plshades.  But for this call we specify
    # non-zero values so that contours are drawn.
    w.plshades(zz, shedge, fill_width, 2, 3.0, 0, w.pltr2, xg2, yg2)

    # Smaller text
    w.plschr(0.0, 0.75 )
    # Small ticks on the vertical axis
    w.plsmaj( 0.0, 0.5 )
    w.plsmin( 0.0, 0.5 )

    num_values[0] = NS + 1
    values[0] = shedge
    (colorbar_width, colorbar_height) = w.plcolorbar ( w.PL_COLORBAR_SHADE | w.PL_COLORBAR_SHADE_LABEL, 0, 0.005, 0.0, 0.0375, 0.875, 0, 1, 1, 0.0, 0.0, 2, 3.0, label_opts, labels, axis_opts, axis_ticks, axis_subticks, num_values, values)

    # Reset text and tick sizes
    w.plschr( 0.0, 1.0 )
    w.plsmaj( 0.0, 1.0 )
    w.plsmin( 0.0, 1.0 )

    w.plcol0(1)
    w.plbox( "bcnst", 0.0, 0, "bcnstv", 0.0, 0 )
    w.plcol0(2)

    # w.plcont(ww, clevel, "pltr2", xg2, yg2, 0)
    w.pllab( "distance", "altitude", "Bogon density" )

    # Example with polar coordinates that demonstrates python wrap support.

    w.plspal0("cmap0_black_on_white.pal")
    w.plspal1("cmap1_gray.pal",1)
    w.plscmap0n(3)

    w.pladv(0)
    w.plvpor(0.1, 0.9, 0.1, 0.9)
    w.plwind(-1.0, 1.0, -1.0, 1.0)

    # Build new coordinate matrices.

    r = arange(NX)/(NX-1.)
    r.shape = (-1,1)
    t = (2.*pi/(NY-1.))*arange(NY-1)
    xg = r*cos(t)
    yg = r*sin(t)
    z = exp(-r*r)*cos(5.*pi*r)*cos(5.*t)

    # Need a new shedge to go along with the new data set.

    zmin = min(z.flat)
    zmax = max(z.flat)
    shedge = zmin + ((zmax - zmin)/NS) * (arange(NS+1))

    w.plpsty(0)

    # Now we can shade the interior region.  Use wrap=2 to simulate additional
    # point at t = 2 pi.
    w.plshades(z, shedge, fill_width, 0, w.pltr2, xg, yg, 2)

    # Smaller text
    w.plschr(0.0, 0.75 )
    # Small ticks on the vertical axis
    w.plsmaj( 0.0, 0.5 )
    w.plsmin( 0.0, 0.5 )

    num_values[0] = NS + 1
    values[0] = shedge
    (colorbar_width, colorbar_height) = w.plcolorbar ( w.PL_COLORBAR_SHADE | w.PL_COLORBAR_SHADE_LABEL, 0, 0.005, 0.0, 0.0375, 0.875, 0, 1, 1, 0.0, 0.0, cont_color, cont_width, label_opts, labels, axis_opts, axis_ticks, axis_subticks, num_values, values)

    # Reset text and tick sizes
    w.plschr( 0.0, 1.0 )
    w.plsmaj( 0.0, 1.0 )
    w.plsmin( 0.0, 1.0 )

    # Now we can draw the perimeter.  (If do before, w.plshades may overlap.)
    t = 2.*pi*arange(PERIMETERPTS)/(PERIMETERPTS-1.)
    px = cos(t)
    py = sin(t)


    w.plcol0(1)
    w.plline( px, py )

    # And label the plot.

    w.plcol0(2)
    w.pllab( "", "",  "Tokamak Bogon Instability" )

    # Restore defaults
    w.plschr( 0.0, 1.0 )
    # cmap0 default color palette.
    w.plspal0("cmap0_default.pal")
    # cmap1 default color palette.
    w.plspal1("cmap1_default.pal",1)

    # Must be done independently because otherwise this changes output files
    # and destroys agreement with C examples.
    #w.plcol0(1)
